File: MiscDemo/app/src/main/java/com/example/miscdemo/MainActivity.java
package com.example.miscdemo;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

public class MainActivity extends AppCompatActivity {

    Button btnMapsAct, btnBroadcastAct, btnServiceAct;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        btnMapsAct = findViewById(R.id.btnMapsActivity);
        btnBroadcastAct = findViewById(R.id.btnBroadcastAct);
        btnServiceAct = findViewById(R.id.btnServiceAct);

        btnServiceAct.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(MainActivity.this, ServiceActivity.class));
            }
        });

        btnMapsAct.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(MainActivity.this, MapsActivity.class));
            }
        });

        btnBroadcastAct.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(MainActivity.this, BroadcaastActivity.class));
            }
        });
    }
}

File: MemeShare/app/src/main/java/com/example/memeshare/MainActivity.kt
package com.example.memeshare

import android.content.Intent
import android.graphics.drawable.Drawable
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.Toast
import com.android.volley.Request
import com.android.volley.Response
import com.android.volley.cronet.CronetHttpStack
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    var currentUrl: String? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        shareButton.setOnClickListener(View.OnClickListener {
            val intent = Intent(Intent.ACTION_SEND)
            intent.type = "text/plain"
            intent.putExtra(Intent.EXTRA_TEXT, "Check the meme from reddit $currentUrl")
            val chooser = Intent.createChooser(intent, "Share this meme using....")
            startActivity(chooser)

        })

        nextButton.setOnClickListener(View.OnClickListener {
            loadMeme()
        })

        loadMeme()

    }

    private fun loadMeme(){
        // Instantiate the RequestQueue.

        val url = "https://meme-api.herokuapp.com/gimme"

        progress_circular.visibility = View.VISIBLE

        // Request a string response from the provided URL.
        // Request a string response from the provided URL.
        val jsonObjectRequest = JsonObjectRequest(
                Request.Method.GET, url, null,
                { response ->
                    currentUrl = response.getString("url")

                    Glide.with(this).load(currentUrl).listener(object : RequestListener<Drawable>{
                        override fun onResourceReady(
                                resource: Drawable?,
                                model: Any?,
                                target: Target<Drawable>?,
                                dataSource: DataSource?,
                                isFirstResource: Boolean
                        ): Boolean {
                            progress_circular.visibility = View.GONE
                            return false
                        }

                        override fun onLoadFailed(
                                e: GlideException?,
                                model: Any?,
                                target: Target<Drawable>?,
                                isFirstResource: Boolean
                        ): Boolean {
                            progress_circular.visibility = View.GONE
                            return false
                        }
                    }).into(imageView)
                },
                {
                    progress_circular.visibility = View.GONE
                    Toast.makeText(this, "Something went wrong", Toast.LENGTH_LONG).show()
                })

        // Add the request to the RequestQueue.
        MySingleton.getInstance(this).addToRequestQueue(jsonObjectRequest)
    }
}